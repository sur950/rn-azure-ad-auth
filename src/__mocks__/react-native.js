import Linking from './linking'
import AzureAuthNative from './azure-auth'
import AsyncStorage from './async-storage'

const mock = {}

mock.Linking = new Linking()
mock.NativeModules = { AzureAuth: new AzureAuthNative() }
mock.Platform = {OS: 'SomeMobile'}
mock.AsyncStorage = new AsyncStorage()

module.exports = mock