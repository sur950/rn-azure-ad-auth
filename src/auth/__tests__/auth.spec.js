import Auth from '../'
import fetchMock from 'fetch-mock'
jest.mock('react-native')
jest.mock('@react-native-async-storage/async-storage')
import { NativeModules } from 'react-native'
const AzureAuth = NativeModules.AzureAuth
import Scope from '../../token/scope'

describe('auth', () => {

    const authorityUrl = 'https://login.microsoftonline.com/common/oauth2/v2.0/'
    const clientId = 'A_CLIENT_ID_OF_YOUR_ACCOUNT'
    const redirectUri = 'https://mysite.com/callback'
    const scope = new Scope() // base scope
    const state = 'a random state for auth'
    const verifier = 'some Azure verifier code'

    const tokens = {
        status: 200,
        body: {
            access_token: 'an access token',
            id_token: 'an id token',
            expires_in: 1234567890,
            state,
            scope: 'openid'
        },
        headers: { 'Content-Type': 'application/json' }
    }
    const imageResponse = {
        status: 200,
        blob: {},
        headers: { 'Content-Type': 'image/png'}
    }
    const oauthError = {
        status: 400,
        body: {
            error: 'invalid_request',
            error_description: 'Invalid grant'
        },
        headers: { 'Content-Type': 'application/json' }
    }
    const unexpectedError = {
        status: 500,
        body: 'Internal Server Error....',
        headers: { 'Content-Type': 'text/plain' }
    }
    const auth  = new Auth({authorityUrl, clientId})

    beforeEach(() => {
        beforeEach(() => {
            NativeModules.AzureAuth = AzureAuth
            AzureAuth.reset()
        })
        fetchMock.restore()
    })

    describe('constructor', () => {
        it('should build with authority and Client ID', () => {
            const auth = new Auth({authorityUrl, clientId})
            expect(auth.clientId).toEqual(clientId)
        })

        it('should fail without clientId', () => {
            expect(() => new Auth({authorityUrl})).toThrowErrorMatchingSnapshot()
        })

        it('should fail without authority', () => {
            expect(() => new Auth({clientId})).toThrowErrorMatchingSnapshot()
        })
    })


    describe('loginUrl', () => {
        it('should return default authorize url', () => {
            expect(auth.loginUrl({
                responseType: 'code',
                scope: scope.toString(),
                state: 'a_random_state'
            })).toMatchSnapshot()
        })

        it('should return default authorize url with extra parameters', () => {
            expect(auth.loginUrl({
                responseType: 'code',
                scope: scope.toString(),
                state: 'a_random_state',
                connection: 'facebook'
            })).toMatchSnapshot()
        })

        it('should use redirect url from Auth init params', () => {
            const ruri = new RegExp(`.*redirect_uri=${encodeURIComponent(auth.redirectUri)}.*`)
            expect(auth.loginUrl({
                responseType: 'code',
                redirectUri,
                scope: scope.toString(),
                state: 'a_random_state',
                connection: 'facebook'
            })).toMatch(ruri)
        })

    })

    describe('logoutUrl', () => {
        it('should return default logout url', () => {
            expect(auth.logoutUrl({})).toMatchSnapshot()
        })

        it('should return logout url with extra parameters', () => {
            expect(auth.logoutUrl({
                redirectUri:'https://bing.com'
            })).toMatchSnapshot()
        })

        it('should return logout url with skipping unknown parameters', () => {
            expect(auth.logoutUrl({
                redirectUri:'https://bing.com',
                shouldNotBeThere: 'really'
            })).toMatchSnapshot()
        })
    })

    describe('code exchange', () => {
        it('should send correct payload', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', tokens)
            expect.assertions(1)
            await auth.exchange({code: 'a code', code_verifier: verifier, scope: 'openid'})
            expect(fetchMock.lastCall()).toMatchSnapshot()
        })

        it('should return successful response', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', tokens)
            expect.assertions(1)
            const parameters = {code: 'a code', code_verifier: verifier, scope: 'openid'}
            await expect(auth.exchange(parameters)).resolves.toMatchSnapshot()
        })

        it('should handle oauth error', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', oauthError)
            expect.assertions(1)
            const parameters = {code: 'a code', code_verifier: verifier, scope: 'openid'}
            await expect(auth.exchange(parameters)).rejects.toMatchSnapshot()
        })

        it('should handle unexpected error', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', unexpectedError)
            expect.assertions(1)
            const parameters = {code: 'a code', code_verifier: verifier, scope: 'openid'}
            await expect(auth.exchange(parameters)).rejects.toMatchSnapshot()
        })
    })

    describe('refresh token', () => {
        const parameters = {refreshToken: 'a refresh token of a user', scope: 'openid'}
        it('should send correct payload', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', tokens)
            expect.assertions(1)
            await auth.refreshTokens(parameters)
            expect(fetchMock.lastCall()).toMatchSnapshot()
        })

        it('should return successful response', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', tokens)
            expect.assertions(1)
            await expect(auth.refreshTokens(parameters)).resolves.toMatchSnapshot()
        })

        it('should handle oauth error', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', oauthError)
            expect.assertions(1)
            await expect(auth.refreshTokens(parameters)).rejects.toMatchSnapshot()
        })

        it('should handle unexpected error', async () => {
            fetchMock.postOnce('https://login.microsoftonline.com/common/oauth2/v2.0/token', unexpectedError)
            expect.assertions(1)
            await expect(auth.refreshTokens(parameters)).rejects.toMatchSnapshot()
        })
    })

    describe('graph requests', () => {
        it('should return blob', async () => {
            fetchMock.getOnce('https://graph.microsoft.com/v1.0/me/photo/$value', imageResponse)
            expect.assertions(1)
            await expect(auth.msGraphRequest({token: 'fake_token', path: 'me/photo/$value'})).resolves.toMatchSnapshot()
        })
    })

})