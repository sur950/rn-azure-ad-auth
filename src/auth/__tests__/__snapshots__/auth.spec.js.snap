// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth code exchange should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth code exchange should handle unexpected error 1`] = `[aad.response.invalid: Internal Server Error]`;

exports[`auth code exchange should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth code exchange should send correct payload 1`] = `
Array [
  "https://login.microsoftonline.com/common/oauth2/v2.0/token",
  Object {
    "body": "&code=a%20code&code_verifier=some%20Azure%20verifier%20code&scope=openid&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback&grant_type=authorization_code",
    "headers": Object {
      "Accept": "application/json, image/*;q=0.1",
      "Content-Length": 211,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;

exports[`auth constructor should fail without authority 1`] = `"Missing Azure authority base URL"`;

exports[`auth constructor should fail without clientId 1`] = `"Missing clientId in parameters"`;

exports[`auth graph requests should return blob 1`] = `
Object {
  "blob": Object {},
  "headers": Object {
    "Content-Type": "image/png",
  },
  "status": 200,
}
`;

exports[`auth loginUrl should return default authorize url 1`] = `"https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&scope=offline_access%20openid%20profile&state=a_random_state&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback"`;

exports[`auth loginUrl should return default authorize url with extra parameters 1`] = `"https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&scope=offline_access%20openid%20profile&state=a_random_state&connection=facebook&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback"`;

exports[`auth logoutUrl should return default logout url 1`] = `"https://login.microsoftonline.com/common/oauth2/v2.0/logout?post_logout_redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback"`;

exports[`auth logoutUrl should return logout url with extra parameters 1`] = `"https://login.microsoftonline.com/common/oauth2/v2.0/logout?post_logout_redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback"`;

exports[`auth logoutUrl should return logout url with skipping unknown parameters 1`] = `"https://login.microsoftonline.com/common/oauth2/v2.0/logout?post_logout_redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback"`;

exports[`auth refresh token should handle oauth error 1`] = `null`;

exports[`auth refresh token should handle unexpected error 1`] = `null`;

exports[`auth refresh token should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth refresh token should send correct payload 1`] = `
Array [
  "https://login.microsoftonline.com/common/oauth2/v2.0/token",
  Object {
    "body": "&refresh_token=a%20refresh%20token%20of%20a%20user&scope=offline_access%20openid%20profile&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&grant_type=refresh_token&redirect_uri=my_bundle_id%3A%2F%2Fmy_bundle_id%2FSomeMobile%2Fcallback",
    "headers": Object {
      "Accept": "application/json, image/*;q=0.1",
      "Content-Length": 224,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;
